### ISSUE TITLE
Implement a Cross-Modal Meta-Ensemble Stacking GNN, LSTM, Tabular, and Benford Signals

### ISSUE THREAD

**Issue by Inkman007** [Ledger-Lenz/Ledgerlens-core#316]
## Summary

LedgerLens currently fuses its four detection modalities — the tabular RF/XGBoost/LightGBM ensemble, the GNN ring score, the temporal sequence score, and raw Benford metrics — through three different, independently-tuned mechanisms scattered across the codebase: GNN and Benford scores are fed in as *input features* to the tabular ensemble, the temporal sequence score is blended post-hoc via a fixed linear weight (`temporal_weight`), and Benford is *also* blended a second time inside `RiskScore.combine`'s `benford_component`. This issue replaces that ad-hoc, partially-duplicated fusion with one principled cross-modal meta-ensemble: a calibrated meta-learner that stacks the tabular ensemble's output, the GNN ring score, the temporal score, and Benford metrics as peer signals, with explicit per-modality availability handling.

## Background & Context

`docs/ensemble_stacking.md` documents an existing stacking layer, but it operates *within* one modality: RF, XGBoost, and LightGBM's out-of-fold predictions are combined by a logistic-regression meta-learner (`models/meta_learner.joblib`) into one calibrated tabular probability. That's the layer this issue builds *on top of*, not the layer it replaces.

Looking at `run_pipeline.py`'s scoring loop concretely: `gnn_scores` (from `detection/gnn_ring_detector.py`) is passed into `build_feature_vector(...)` as one of the inputs that produces `GNN_FEATURE_NAMES` values, which become ordinary columns the tabular ensemble sees — i.e. GNN fusion already happens, but only implicitly, diluted among 35+ other features with no separate accounting for how much the GNN signal specifically moved the final score. Benford metrics are treated the same way (raw features in `BENFORD_FEATURE_NAMES` feed the tabular ensemble) *and* separately re-enter through `RiskScore.combine`'s explicit `benford_component = min(benford_mad / benford_mad_threshold, 1.0) * 100`, which is blended `0.3 * benford_component + 0.7 * ml_component` — meaning Benford's influence on the final score is a mix of "however much the tree ensemble weighted the Benford feature columns" plus a fixed, hand-set 30% linear weight, compounded. The temporal score is fused differently again: `run_pipeline.py`'s `adjust_score_with_temporal()` calls `detection.risk_score.temporal_risk_adjustment(snapshot_score, temporal_score, history_days, temporal_weight=settings.temporal_weight)`, a simple linear blend (`0.7 * snapshot + 0.3 * temporal`, with `temporal_weight` default `0.3`) applied to the score *after* `RiskScore.combine` has already run — a genuinely separate, non-learned fusion step with its own hardcoded-by-config weight. And there's a second, distinct temporal signal — `detection/temporal_patterns.py`'s `temporal_anomaly_score` (ARIMA residual + LSTM autoencoder + IAT variance + burst synchrony, per that module's `TemporalPatternScorer`) — which is yet another tabular feature column, unrelated to the `temporal_model.py` LSTM/Transformer sequence score that drives `adjust_score_with_temporal`.

The result: three genuinely different fusion strategies (feature-level dilution, fixed post-hoc linear blend, double-counted explicit blend) for what should be one coherent question — "given everything each modality independently believes, what's the calibrated probability this wallet is engaged in wash trading?" A cross-modal meta-ensemble, structured exactly like the existing intra-tabular stacking layer in `docs/ensemble_stacking.md` (walk-forward OOF generation, logistic-regression or gradient-boosted meta-learner, disagreement features) but one level up — stacking *modality-level* scores instead of *base-model-level* predictions — replaces all three ad-hoc mechanisms with one learned, auditable one, and gives an explicit answer to "was this wallet flagged primarily because of the GNN, the temporal pattern, or Benford" the same way `docs/ensemble_stacking.md`'s meta-learner coefficients already answer "which base model does the ensemble trust."

## Objectives

- [ ] Implement `detection/meta_ensemble.py` with `CrossModalMetaEnsemble`, a second-level stacking meta-learner over four modality scores: `tabular_probability` (existing RF/XGB/LGBM + intra-tabular meta-learner output), `gnn_ring_score` (`detection/gnn_ring_detector.py`), `temporal_score` (unify `temporal_model.py`'s sequence score and `temporal_patterns.py`'s `temporal_anomaly_score` into one input, documented explicitly), and `benford_mad_normalized` (raw Benford signal, not pre-blended)
- [ ] Implement per-modality availability handling: any modality can be missing for a given wallet (e.g. GNN score absent for a wallet not in any detected ring, temporal score absent for wallets with `history_days < 7`) — the meta-learner must handle missing inputs via either (a) trained-in imputation matching each modality's documented "no signal" default (`0.0` for GNN per `GNNRingDetector.predict`'s not-fitted fallback, snapshot-only for temporal per `temporal_risk_adjustment`'s existing `history_days < 7` fallback), or (b) a modality-presence indicator feature, and the design decision must be documented with rationale
- [ ] Train the meta-learner via walk-forward OOF generation, reusing `docs/ensemble_stacking.md`'s existing 5-fold temporal-CV-with-purge-gap methodology applied at the modality level instead of the base-model level
- [ ] Replace `run_pipeline.py`'s `adjust_score_with_temporal()` post-hoc blend and `RiskScore.combine`'s `benford_component` double-counting with a single call into `CrossModalMetaEnsemble.score(...)`, preserving `RiskScore.combine`'s existing bounded-adjustment parameters (`sandwich_weight`, `pdc_discount_weight`) as-is since those are genuinely different signal types (price-manipulation, causal PDC) not covered by this stacking layer
- [ ] Add a `GET /scores/{wallet}/modality-breakdown` endpoint (or extend the existing `/scores/{wallet}/explain` SHAP endpoint) surfacing each modality's raw score, its meta-learner coefficient, and its contribution to the final composite — the cross-modal analog of `docs/ensemble_stacking.md`'s logged meta-learner coefficients
- [ ] Add a fallback path: when `models/cross_modal_meta_learner.joblib` is absent, fall back to the current fixed-weight behaviour exactly (temporal linear blend + Benford double-blend), logging `INFO cross-modal meta-learner not found; using legacy fixed-weight fusion`, mirroring `docs/ensemble_stacking.md`'s existing "Fallback Behaviour" pattern for backward compatibility
- [ ] Benchmark AUC-ROC / AUC-PR of the meta-ensemble against the current fixed-weight fusion on the existing labelled synthetic dataset and the backtest framework (`backtesting/backtest_runner.py`)

## Technical Requirements

**Meta-ensemble interface** (`detection/meta_ensemble.py`):

```python
@dataclass
class ModalityScores:
    tabular_probability: float                 # from ensemble_reweighter / stacking meta-learner
    gnn_ring_score: float | None                # None when GNNRingDetector has no signal for this wallet
    temporal_score: float | None                # None when history_days < 7
    benford_mad_normalized: float               # min(benford_mad / benford_mad_threshold, 1.0), always available

class CrossModalMetaEnsemble:
    def __init__(self, model_path: str = "models/cross_modal_meta_learner.joblib"): ...

    def score(self, modality_scores: ModalityScores) -> "MetaEnsembleResult": ...

@dataclass
class MetaEnsembleResult:
    probability: float                          # calibrated final probability, 0-1
    contributions: dict[str, float]              # {"tabular": 0.52, "gnn": 0.21, "temporal": 0.14, "benford": 0.13}
    modalities_available: list[str]              # which inputs were non-null for this wallet
```

**Missing-modality feature encoding** — following the "disagreement features" precedent in `docs/ensemble_stacking.md`, add explicit availability indicators to the meta-learner's input vector rather than silently imputing zero (which would bias the learner toward interpreting "no ring detected" identically to "definitely not in a ring"):

```python
def build_meta_features(scores: ModalityScores) -> np.ndarray:
    return np.array([
        scores.tabular_probability,
        scores.gnn_ring_score if scores.gnn_ring_score is not None else 0.0,
        1.0 if scores.gnn_ring_score is not None else 0.0,          # gnn_available
        scores.temporal_score if scores.temporal_score is not None else 0.0,
        1.0 if scores.temporal_score is not None else 0.0,          # temporal_available
        scores.benford_mad_normalized,
    ])
```

**Integration point** — `run_pipeline.py`'s scoring loop, replacing the two-step `RiskScore.combine()` + `adjust_score_with_temporal()` sequence:

```python
modality_scores = ModalityScores(
    tabular_probability=probability,
    gnn_ring_score=gnn_scores.get(account),
    temporal_score=temporal_prob if history_days >= 7 else None,
    benford_mad_normalized=min(features.get("benford_mad_24h", 0.0) / settings.benford_mad_threshold, 1.0),
)
meta_result = meta_ensemble.score(modality_scores)
score = RiskScore.combine(
    wallet=account, asset_pair=pair_key,
    benford_mad=features.get("benford_mad_24h", 0.0),
    benford_mad_threshold=settings.benford_mad_threshold,
    ml_probability=meta_result.probability,   # meta-ensemble output replaces raw tabular probability here
    ml_confidence=confidence,
    ...
)
```

**Configuration** (`.env.example`):

```
CROSS_MODAL_META_ENSEMBLE_ENABLED=false
CROSS_MODAL_META_LEARNER_PATH=models/cross_modal_meta_learner.joblib
CROSS_MODAL_MIN_TRAINING_SAMPLES=500
```

## Security Considerations

- `GET /scores/{wallet}/modality-breakdown` exposes which detection modality drove a score — same interpretability-vs-evasion tradeoff already accepted for SHAP (`/scores/{wallet}/explain`) and causal explanations (`/scores/{wallet}/causal-explanation`); gate it identically (no new exposure beyond what those endpoints already establish as acceptable)
- Meta-learner model file (`cross_modal_meta_learner.joblib`) is loaded via `joblib.load` — apply the same integrity discipline already used for `.pt`/SHA-256-checked model artifacts elsewhere in the codebase if this repo's `joblib` loading path doesn't already have one; do not introduce an unchecked deserialization path
- Missing-modality indicator features must not become an evasion vector: an adversary strategically avoiding ring-detectable structure to keep `gnn_ring_score=None` should not systematically lower their score more than a wallet that *was* evaluated by the GNN and scored low — validate this empirically in the benchmark (Objective 7) and document the finding either way

## Testing Requirements

- `tests/test_meta_ensemble.py`
- Test: `build_meta_features` produces correct availability indicators for all four combinations of GNN/temporal presence
- Test: `CrossModalMetaEnsemble.score()` with all four modalities present matches expected output for a fixed trained fixture model
- Test: `CrossModalMetaEnsemble.score()` with GNN and temporal both `None` (new wallet, no ring) still produces a valid probability using tabular + Benford only
- Test: fallback path — `models/cross_modal_meta_learner.joblib` absent — reproduces the exact current fixed-weight (`temporal_risk_adjustment` + `RiskScore.combine` double-blend) output for a set of golden fixtures, guaranteeing no silent behaviour change for deployments that don't opt in
- Test: `run_pipeline.py` integration — `CROSS_MODAL_META_ENSEMBLE_ENABLED=true` produces scores within `[0, 100]` and calls `meta_ensemble.score()` exactly once per wallet/asset-pair
- Test: `GET /scores/{wallet}/modality-breakdown` returns contributions summing to the final probability within floating-point tolerance
- Backtest comparison test (`backtesting/backtest_runner.py`): meta-ensemble AUC-ROC/AUC-PR reported alongside legacy fixed-weight fusion on the same labelled dataset

## Documentation Requirements

- New `docs/cross_modal_meta_ensemble.md`: motivation (the three-mechanism fragmentation this replaces, with the specific `run_pipeline.py`/`risk_score.py` line-level references above), architecture, missing-modality handling rationale, training procedure, fallback behaviour, and a worked example of `modality-breakdown` output
- Update `docs/ensemble_stacking.md` to cross-reference this as "level 2" stacking, clarifying it operates on modality scores, not base-model predictions
- Update README.md's "Machine Learning Layer" and "Interpretability" sections to mention the cross-modal meta-ensemble
- Update `.env.example` with the new `CROSS_MODAL_*` variables

## Definition of Done

- [ ] All objectives completed
- [ ] Tests pass (`pytest`)
- [ ] No regressions on existing test suite
- [ ] PR reviewed and approved

## For Contributors

**When applying for this issue, please specify:**
- Your area of specialty (e.g., ensemble methods, model stacking, multi-modal fusion)
- Relevant experience with: stacking/meta-learning architectures, handling missing modalities in ensemble models, calibration (Platt scaling, isotonic regression)
- Your approach or initial thoughts on the implementation
- Estimated time to complete

**Ideal contributor profile:** ML engineer with hands-on multi-modal or multi-signal ensemble stacking experience, who has dealt with missing-modality handling in a production scoring system before and can reason carefully about the fallback/backward-compatibility requirements this issue imposes on a live detection pipeline.

**Comment by Inkman007** [comment]
Closing as part of a full issue-tracker reset: the project's open issues have been replaced with a new curated set of 125 advanced-level issues covering the codebase end to end. No action needed here; if this item is still relevant, please open a fresh issue referencing it.
